src/main/java/com/google/cloud/tools/app/cloudsdk/CloudSdkAppEngineFactory.java

```diff
@@ -0,0 +1,216 @@
+    return new CloudSdkAppEngineFlexibleStaging();
+  }
+
+  @Override
+  public GenConfigUtility genConfigUtility() {
+    return new CloudSdkAppEngineGenConfig(cloudSdk);
+  }
+
+  @Override
+  public AppEngineInstances instances() {
+    return new CloudSdkAppEngineInstances(cloudSdk);
+  }
+
+  @Override
+  public AppEngineLogs logs() {
+    return new CloudSdkAppEngineLogs(cloudSdk);
+  }
+
+  @Override
+  public AppEngineServices services() {
+    return new CloudSdkAppEngineServices(cloudSdk);
+  }
+
+  @Override
+  public AppEngineStandardStaging standardStaging() {
+    return new CloudSdkAppEngineStandardStaging(cloudSdk);
+  }
+
+  @Override
+  public AppEngineVersions versions() {
+    return new CloudSdkAppEngineVersions(cloudSdk);
+  }
+
+  public static class Builder {
+    private CloudSdk.Builder cloudSdkBuilder;
+
+    /**
+     * The home directory of Google Cloud SDK. If not set, will attempt to look for the SDK in known
+     * install locations.
+     */
+    public Builder sdkPath(File sdkPathFile) {
+      cloudSdkBuilder.sdkPath(sdkPathFile);
+      return this;
+    }
+
+    /**
+     * The metrics environment.
+     */
+    public Builder appCommandMetricsEnvironment(String appCommandMetricsEnvironment) {
+      cloudSdkBuilder.appCommandMetricsEnvironment(appCommandMetricsEnvironment);
+      return this;
+    }
+
+    /**
+     * The metrics environment version.
+     */
+    public Builder appCommandMetricsEnvironmentVersion(String appCommandMetricsEnvironmentVersion) {
+      cloudSdkBuilder.appCommandMetricsEnvironmentVersion(appCommandMetricsEnvironmentVersion);
+      return this;
+    }
+
+    /**
+     * Configures usage of gsutil.
+     */
+    public Builder appCommandGsUtil(Integer appCommandGsUtil) {
+      cloudSdkBuilder.appCommandGsUtil(appCommandGsUtil);
+      return this;
+    }
+
+    /**
+     * Sets the path the credential override file.
+     */
+    public Builder appCommandCredentialFile(File appCommandCredentialFile) {
+      cloudSdkBuilder.appCommandCredentialFile(appCommandCredentialFile);
+      return this;
+    }
+
+    /**
+     * Sets the format for printing command output resources. The default is a command-specific
+     * human-friendly output format. The supported formats are: csv, default, flattened, json, list,
+     * multi, none, table, text, value, yaml. For more details run $ gcloud topic formats.
+     */
+    public Builder appCommandOutputFormat(String appCommandOutputFormat) {
+      cloudSdkBuilder.appCommandOutputFormat(appCommandOutputFormat);
+      return this;
+    }
+
+    /**
+     * Whether to run commands asynchronously.
+     */
+    public Builder async(boolean async) {
+      cloudSdkBuilder.async(async);
+      return this;
+    }
+
+    /**
+     * Adds a client consumer of process standard output. If none, output will be inherited by
+     * parent process.
+     */
+    public Builder addStdOutLineListener(ProcessOutputLineListener stdOutLineListener) {
```

Comment:
this actually reminds me that we should remove the `add` prefix and use varargs instead